### ISSUE TITLE
Segmentation fault in index_reader

### ISSUE THREAD

**Issue by pdziepak** [scylladb/scylla#2696]
*Installation details*
Scylla version (or git commit hash): 2.0-rc1
Cluster of 3, RF=3.

Scylla was populated using scylla-bench:
```
scylla-bench -duration 1h -partition-count 1000000 -clustering-row-count 1000 -clustering-row-size 128 -connection-count 32 -measure-latency=false -replication-factor 3 -concurrency=256 -workload=uniform -mode=write
```

Then the following read workload caused segmentation fault on two nodes almost immediately:
```
scylla-bench -duration 15m -partition-count 1000000 -clustering-row-count 1000 -clustering-row-size 128 -connection-count 32 -measure-latency=false -replication-factor 3 -concurrency=256 -workload=uniform -mode=read
```

Backtrace:
```
segfault at 2004290 ip 0000000000cc8b62 sp 00007f8996df4fa0 error 7 in scylla[400000+248b000]
Segmentation fault on shard 29.
0x00000000004eef08
0x00000000004ef025
0x00000000004ef073
0x00007f89afff50ff
0x0000000000cc8b61
0x0000000000ccac93
0x00000000004eccbf
0x00000000005407aa
0x0000000000542050
0x000000000059a36d
0x00007f89affeddc4
0x00007f89afd1acec
seastar::backtrace_buffer::append_backtrace() at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:277
 (inlined by) print_with_backtrace at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:288
print_with_backtrace at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:295
sigsegv_action at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:3508
 (inlined by) operator() at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:3494
 (inlined by) _FUN at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:3490
?? ??:0
_ZN7seastar13lw_shared_ptrIN8sstables7sstableEEC4ERKS3_ at /usr/src/debug/scylla-2.0.rc1/seastar/core/shared_ptr.hh:178
 (inlined by) void std::experimental::fundamentals_v1::_Optional_base<sstables::index_reader::reader, true>::_M_construct<seastar::lw_shared_ptr<sstables::sstable>&, seastar::io_priority_class const&, unsigned long const&, unsigned long const&, unsigned long const&>(seastar::lw_shared_ptr<sstables::sstable>&, seastar::io_priority_class const&, unsigned long const&, unsigned long const&, unsigned long const&) at /opt/scylladb/include/c++/5.3.1/experimental/optional:308
 (inlined by) void std::experimental::fundamentals_v1::optional<sstables::index_reader::reader>::emplace<seastar::lw_shared_ptr<sstables::sstable>&, seastar::io_priority_class const&, unsigned long const&, unsigned long const&, unsigned long const&>(seastar::lw_shared_ptr<sstables::sstable>&, seastar::io_priority_class const&, unsigned long const&, unsigned long const&, unsigned long const&) at /opt/scylladb/include/c++/5.3.1/experimental/optional:531
 (inlined by) _ZZZN8sstables12index_reader15advance_to_pageEmENKUlmE_clEmENKUlOT_E_clIN7seastar6futureIJEEEEEDaS3_ at /usr/src/debug/scylla-2.0.rc1/sstables/index_reader.hh:253
_ZN7seastar8futurizeINS_6futureIJSt6vectorIN8sstables11index_entryESaIS4_EEEEEE5applyIZZNS3_12index_reader15advance_to_pageEmENKUlmE_clEmEUlOT_E_JNS1_IJEEEEEES7_SD_DpOT0_ at /usr/src/debug/scylla-2.0.rc1/seastar/core/future.hh:1312
 (inlined by) _ZZN7seastar6futureIJEE12then_wrappedIZZN8sstables12index_reader15advance_to_pageEmENKUlmE_clEmEUlOT_E_NS0_IJSt6vectorINS3_11index_entryESaISA_EEEEEEET0_S7_ENUlS7_E_clINS_12future_stateIJEEEEEDaS7_ at /usr/src/debug/scylla-2.0.rc1/seastar/core/future.hh:940
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedIZZN8sstables12index_reader15advance_to_pageEmENKUlmE_clEmEUlOT_E_NS1_IJSt6vectorINS4_11index_entryESaISB_EEEEEEET0_S8_EUlS8_E_JEE3runEv at /usr/src/debug/scylla-2.0.rc1/seastar/core/future.hh:395
seastar::reactor::run_tasks(seastar::circular_buffer<std::unique_ptr<seastar::task, std::default_delete<seastar::task> >, std::allocator<std::unique_ptr<seastar::task, std::default_delete<seastar::task> > > >&) at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:2316
seastar::reactor::run() at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:2774
operator() at /usr/src/debug/scylla-2.0.rc1/seastar/core/reactor.cc:3707
std::function<void ()>::operator()() const at /opt/scylladb/include/c++/5.3.1/functional:2271
 (inlined by) seastar::posix_thread::start_routine(void*) at /usr/src/debug/scylla-2.0.rc1/seastar/core/posix.cc:52
?? ??:0
?? ??:0
```

**Comment by pdziepak** [comment]
The instruction that segfaults is:
```
addq   $0x1,0x0(%rbp)
```
`%rbp` is:
```
0x2004290 <vtable for seastar::shared_ptr_count_for<service::digest_read_resolver>+16>
```

Looks like some kind of invalid memory access, either `index_reader` is used after free or something else is corrupting it.

**Comment by avikivity** [comment]
Try debug mode -- I had very similar errors when something was captured by value where it should have been captured by reference.

**Comment by pdziepak** [comment]
I'm running smaller version of this workload locally in the debug mode for some time now, but no luck so far. Debug mode is much slower and orders continuations differently so it may not be easy to reproduce.

**Comment by pdziepak** [comment]
One potential cause of this is lack of exception safety in the code using `index_reader` (e.g. scylladb/seastar#135). I guess I need to resurrect memory allocation failure injector to hunt for those problems.

I am not fully convinced that this was the real cause, most of allocations which failure could result in a segmentation fault like this one should be relatively small and if we cannot satisfy them then it is likely that we would not be able to allocate memory for new continuations, something that is deadly to Scylla anyway.

**Comment by tzach** [comment]
@pdziepak ping

**Comment by pdziepak** [comment]
I haven't been able to reproduce this issue. AddressSanitizer didn't help either.
Memory allocation fault injector did show several problems, however, for which I'm going to send patches soon, but it is unknown if any of them solves the rely cause of this segmentation fault.
There is also a libstdc++ bug [1] that was found.

[1] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81891

**Comment by avikivity** [comment]
Nice to see that gcc is so responsive.

We plan to switch to self-built gcc, so we can carry the patch ourselves 
as well.

**Comment by pdziepak** [comment]
Patches fixing exception safety have been merged (93eaa95378da53d925c2f57e24b958dc90c68a07) and backported (701128f8a1c54a6d51701d476a873de9f8b9339e). There was an issue with `index_reader` related code so there is some hope that the problem was fixed, but considering how hard it was to reproduce the segmentation fault I don't think it will be possible to verify with full certainty that it is now solved.

**Comment by slivne** [comment]
lets close the issue and reopen if it happens.

we can try "hard" to see if it reproduces - e.g. if it took 10 runs to trigger the issue - lets run 100 and if it passes assume its clear.

@pdziepak can you provide info how you triggered the issue

**Comment by pdziepak** [comment]
Well, it is in the first post. First, I populated scylla with the following s-b load:
```
scylla-bench -duration 1h -partition-count 1000000 -clustering-row-count 1000 -clustering-row-size 128 -connection-count 32 -measure-latency=false -replication-factor 3 -concurrency=256 -workload=uniform -mode=write
```

And then started a read workload (basically the same command but `-mode=read`) and the two nodes crashed very quickly (i.e. the duration of the read workload doesn't really matter). This has happened only the first time I've run a workload like this. I've tried running it twice later without exception safety fixes and run it also ~4 times with exception safety fixes. It was not reproduced.
